sir_model.py

```diff
@@ -24,10 +24,9 @@ def get_infection_probas(states, transmissions):
     """
     infected = (states == 1)
     N = len(states)
-    infection_probas = np.zeros(N)
-    for i in range(N):
-        rates = transmissions.toarray()[i][infected == True]
-        infection_probas[i] = 1 - np.prod(1 - rates)
+    infected_transmissions = transmissions.multiply(infected)  # element-wise multiplication for masking
+    infection_probas = (1 - (1-infected_transmissions.toarray()).prod(axis=1)).squeeze()
```

Comment:
could you use (and test if it works) the trick prod_j (1 - A_ij) = exp(sum_j ln(1-A_ij))  ? that way we can keep a sparse matrix all along, here A_ij = infected_transmissions is sparse and so should be ln(1-A_ij)

Reply:
nice, there is even a log1p method for sparse matrices

Reply:
Thanks for the trick! So I have implemented `infection_probas = 1 - np.exp(np.sum(np.log1p(infected_transmissions.multiply(-1)), axis=1))`. It returns the same output as my previous implementation, and I ran it through the Test_inference.ipynb notebook.